src/twisted/internet/defer.py

```diff
@@ -1435,40 +1467,74 @@ def gotResult(r):
                     waiting[0] = False
                     waiting[1] = r
                 else:
-                    _inlineCallbacks(r, g, deferred)
+                    # We are not waiting for deferred result any more
+                    _inlineCallbacks(r, g, status)
 
             result.addBoth(gotResult)
             if waiting[0]:
                 # Haven't called back yet, set flag so that we get reinvoked
                 # and return from the loop
                 waiting[0] = False
-                return deferred
+                status.waitingOn = result
+                return
 
             result = waiting[1]
             # Reset waiting to initial values for next loop.  gotResult uses
             # waiting, but this isn't a problem because gotResult is only
             # executed once, and if it hasn't been executed yet, the return
             # branch above would have been taken.
 
-
             waiting[0] = True
             waiting[1] = None
 
 
+
+def _cancellableInlineCallbacks(g):
+    """
+    Make an C{@}L{inlineCallbacks} cancellable.
+
+    @param g: a generator object returned by calling a function or method
+        decorated with C{@}L{inlineCallbacks}
+
+    @return: L{Deferred} for the C{@}L{inlineCallbacks} that is cancellable.
+    """
+    def cancel(it):
+        it.callbacks, tmp = [], it.callbacks
+        it.addErrback(handleCancel)
+        it.callbacks.extend(tmp)
+        it.errback(_PeculiarError())
+    deferred = Deferred(cancel)
+    status = _CancellationStatus(deferred)
+    def handleCancel(result):
+        result.trap(_PeculiarError)
+        status.deferred = Deferred(cancel)
```

Comment:
The control flow here is hard enough to follow that `handleCancel` definitely deserves a comment explaining it.  Here's my attempt:

Calling `cancel` on the `Deferred` returned by an `inlineCallbacks` decorated function only has an effect if that `Deferred` has no result.  That in turn is only possible if the decorated function has `yield`ed a `Deferred` that itself has no result yet.  In that case, [`waiting[0]` will be `True` after `inlineCallbacks` attaches `gotResult`, and the `_CancellationStatus` instance's `waitingOn` will be the `yield`ed result-less `Deferred`](https://github.com/altendky/twisted/blob/b021148a1bf4d626b22ccf2b18a253a77b5edb42/src/twisted/internet/defer.py#L1463-L1479).

Cancelling `waitingOn` propagates cancellation to the `yield`ed `Deferred` as intended.  This cancellation either suspends `waitingOn`'s callback processing by delegating execution to another result-less `Deferred` or executes `gotResult` and thus resumes the decorated generator.  Now, the generator might emit a final object that `_inlineCallbacks` has to communicate to the caller, but `_inlineCallbacks` can't [`callback` or `errback` its `_CancellationStatus` instance's `deferred`](https://github.com/altendky/twisted/blob/b021148a1bf4d626b22ccf2b18a253a77b5edb42/src/twisted/internet/defer.py#L1417-L1460) because it was just cancelled.  The solution is to replace the cancelled `deferred` attribute with a _new_ `Deferred` whose cancellation function is `cancel` that's also returned into `waitingOn`'s callback chain.  The generator then resumes with all the degrees of freedom it had prior to its `Deferred`'s cancellation; it can return a value or raise an exception that callbacks or errbacks `status.deferred`, which resumes the previous `Deferred`'s callback chain; it can even `yield` a `Deferred` without a result that can then be cancelled again.

This explains why `waitingOn`, which `_CancellationStatus.__init__` sets to `None`, is always a `Deferred` by the time this runs, and why `status.deferred` has to be replaced by a new `Deferred` in `handleCancel`.  The reason that `handleCancel` has to be at the beginning of the callback chain of the `Deferred` returned by `inlineCallbacks` is a little more complicated.

A `Deferred` waiting on a `Deferred` pauses its own callback chain until that `Deferred` fires and propagates cancellation downwards: 

```python
d = Deferred()
d.addCallback(lambda _: d.Deferred())
d.addErrback(print)
d.callback(None)  # <Deferred at 0x1234 waiting on Deferred at 0x5678>
d.cancel()  # prints [Failure instance: Traceback (failure with no frames): <class 'twisted.internet.defer.CancelledError'>: ]
```

`inlineCallbacks`, however, contains its own trampoline [so that `yield`ing `Deferred`s with results doesn't overflow the stack.](https://github.com/altendky/twisted/blob/b021148a1bf4d626b22ccf2b18a253a77b5edb42/src/twisted/internet/defer.py#L1399-L1402), which prevents its `Deferred` from directly waiting on those `yield`ed by its generator:

```python
@inlineCallbacks
def f():
    d = Deferred()
    print(d)  # prints: <Deferred at 0x1234>
    yield Deferred()

print(f())  # prints: <Deferred at 0x5678>; note that it isn't waiting on 0x1234!
```

This is indeed the reason cancellation isn't propagated.  Mimicking the exact behavior of `Deferred`s outside of `inlineCallback`s would be difficult; at what point in the callback chain should execution be suspended?  Fortunately the nature of `inlineCallback`s makes the answer easy.  Because it's not possible to add callbacks to the `Deferred` that run before the generator is resumed, propagating cancellation downward should always happen _before_ any callbacks added to the `inlineCallback`s `Deferred`.  That's why `cancel` splices `handleCancel` in at the beginning of `callbacks`.

Reply:
I'm guessing you don't want this whole dissertation in there so I made a (massively) abbreviated addition after re-understanding this code (thanks).  Let me know if you did want a good chunk more of this explanation included.

Reply:
The documentation as you've written it, is, I think, sufficient.  `inlineCallbacks` might deserve its own design document somewhere that makes clear what my comment tried to make clear, but that would be a separate ticket and PR.